include/cuco/detail/static_multimap.inl

```diff
@@ -0,0 +1,1870 @@
+              cuco::make_pair<Key, Value>(std::move(this->get_empty_key_sentinel()),
+                                          std::move(this->get_empty_value_sentinel()));
+          }
+        }
+      }
+    }  // if running
+
+    warp.sync();
+    if (*warp_counter + warp.size() * vector_width() > buffer_size) {
+      flush_output_buffer(warp,
+                          *warp_counter,
+                          probe_output_buffer,
+                          contained_output_buffer,
+                          num_matches,
+                          probe_output_begin,
+                          contained_output_begin);
+      // First lane reset warp-level counter
+      if (warp.thread_rank() == 0) { *warp_counter = 0; }
+    }
+
+    current_slot = next_slot(current_slot);
+  }  // while running
+}
+
+template <typename Key,
+          typename Value,
+          class ProbeSequence,
+          cuda::thread_scope Scope,
+          typename Allocator>
+template <uint32_t cg_size,
+          uint32_t buffer_size,
+          bool is_outer,
+          typename CG,
+          typename atomicT,
+          typename OutputZipIt1,
+          typename OutputZipIt2,
+          typename PairEqual>
+__device__ void
+static_multimap<Key, Value, ProbeSequence, Scope, Allocator>::device_view::pair_retrieve_impl(
+  CG const& g,
+  value_type const& pair,
+  uint32_t* cg_counter,
+  value_type* probe_output_buffer,
+  value_type* contained_output_buffer,
+  atomicT* num_matches,
+  OutputZipIt1 probe_output_begin,
+  OutputZipIt2 contained_output_begin,
+  PairEqual pair_equal) noexcept
+{
+  const uint32_t lane_id = g.thread_rank();
+
+  auto key          = pair.first;
+  auto current_slot = initial_slot(g, key);
+
+  bool running                      = true;
+  [[maybe_unused]] bool found_match = false;
+
+  while (running) {
+    // TODO: Replace reinterpret_cast with atomic ref when possible. The current implementation
+    // is unsafe!
+    static_assert(sizeof(Key) == sizeof(cuda::atomic<Key>));
+    static_assert(sizeof(Value) == sizeof(cuda::atomic<Value>));
+    value_type slot_contents = *reinterpret_cast<value_type const*>(current_slot);
+
+    auto const slot_is_empty =
+      detail::bitwise_compare(slot_contents.first, this->get_empty_key_sentinel());
+    auto const equals = (not slot_is_empty and pair_equal(slot_contents, pair));
+    auto const exists = g.ballot(equals);
+
+    if (exists) {
+      if constexpr (is_outer) { found_match = true; }
+      auto num_matches    = __popc(exists);
+      uint32_t output_idx = *cg_counter;
+      if (equals) {
+        // Each match computes its lane-level offset
+        auto lane_offset                                  = __popc(exists & ((1 << lane_id) - 1));
+        probe_output_buffer[output_idx + lane_offset]     = pair;
+        contained_output_buffer[output_idx + lane_offset] = slot_contents;
+      }
+      if (0 == lane_id) { (*cg_counter) += num_matches; }
```

Comment:
Good question. :)
By doing `auto const exists = g.ballot(equals);`, I assume all lanes in the CG will fall into the same `if (exists)` branch so this should be safe?

Reply:
I agree all threads in the CG would execute the same branch, but my concern is that they are not guaranteed to execute the same command at the same time. If lane0 executes L1415 before lane1 executes L1408, lane1 could get an incorrect updated `output_idx`, instead of the original one.

Reply:
I think @gaohao95 is right. This can be easily fixed if we move the initialization of `output_idx` to above the `ballot`, then there should never be a problem.

Reply:
Thank you both! Done.